File: ios/Classes/ErrorMapping.swift
// ErrorMapping.swift
//
// Translates ESPProvision's domain error types into the typed FlutterError
// vocabulary expected by the Dart-side `mapPlatformException`.
//
// Codes correspond 1:1 to the concrete subtypes of the sealed
// `EspProvisioningException` in `lib/src/exceptions/`:
//
//   - ble_unavailable          → BleUnavailableException
//   - permission_denied        → PermissionDeniedException (details.permission)
//   - device_not_found         → DeviceNotFoundException   (details.deviceId)
//   - pop_invalid              → PopInvalidException
//   - session_failed           → SessionFailedException
//   - wifi_provisioning_failed → WifiProvisioningFailedException
//   - softap_connection_failed → SoftApConnectionException (details.ssid)
//
// Unknown native errors deliberately fall through to `session_failed` so
// callers always receive a typed Dart exception rather than a raw
// PlatformException leaking out of the public API.

import ESPProvision
import Flutter
import Foundation

enum ErrorMapping {

    /// Maps an `ESPSessionError` to the appropriate FlutterError. Behaviour
    /// of `sessionInitError` is security-level aware: under sec1/sec2 the
    /// most common cause is a wrong PoP rejected during the
    /// SRP6a / X25519 handshake, so we map to `pop_invalid` to drive the
    /// "re-prompt for PoP" UX. Under sec0 the same error means a generic
    /// handshake mismatch, so we fall through to `session_failed`.
    static func flutterError(
        forSessionError error: ESPSessionError,
        security: ESPSecurity,
        ssidOrName: String
    ) -> FlutterError {
        switch error {
        case .noPOP:
            return FlutterError(
                code: "pop_invalid",
                message: "Proof-of-Possession was not supplied to the device handshake",
                details: nil)
        case .sessionInitError:
            if security == .secure || security == .secure2 {
                return FlutterError(
                    code: "pop_invalid",
                    message: "Secure session handshake failed — typically a wrong PoP",
                    details: nil)
            }
            return FlutterError(
                code: "session_failed",
                message: error.description,
                details: nil)
        case .securityMismatch, .encryptionError:
            return FlutterError(
                code: "session_failed",
                message: error.description,
                details: nil)
        case .softAPConnectionFailure:
            return FlutterError(
                code: "softap_connection_failed",
                message: error.description,
                details: ["ssid": ssidOrName])
        case .bleFailedToConnect:
            return FlutterError(
                code: "device_not_found",
                message: error.description,
                details: ["deviceId": ssidOrName])
        case .versionInfoError, .sendDataError, .sessionNotEstablished, .noUsername:
            return FlutterError(
                code: "session_failed",
                message: error.description,
                details: nil)
        }
    }

    /// Wi-Fi scan failures are not security-sensitive; map them all to
    /// `session_failed` so the caller can retry the scan without
    /// re-prompting the user for credentials.
    static func flutterError(forWifiScanError error: ESPWiFiScanError) -> FlutterError {
        return FlutterError(
            code: "session_failed",
            message: error.description,
            details: nil)
    }

    /// Maps an `ESPProvisionError` to a `ProvisioningResult` map (delivered
    /// in-band to the Dart side as a success completion). Callers above
    /// should special-case `.wifiStatusDisconnected` themselves and raise
    /// `wifi_provisioning_failed` instead of using this map — the
    /// channel-level breakdown is an exception, not a result.
    static func provisioningResultMap(for error: ESPProvisionError) -> [String: Any?] {
        switch error {
        case .wifiStatusAuthenticationError:
            return [
                "status": "authFailed",
                "rawCode": error.code,
                "rawMessage": error.description,
            ]
        case .wifiStatusNetworkNotFound:
            return [
                "status": "networkNotFound",
                "rawCode": error.code,
                "rawMessage": error.description,
            ]
        case .configurationError(let underlying):
            return [
                "status": "deviceInternalError",
                "rawCode": error.code,
                "rawMessage": "configurationError: \(underlying.localizedDescription)",
            ]
        case .wifiStatusError(let underlying):
            return [
                "status": "unknown",
                "rawCode": error.code,
                "rawMessage": "wifiStatusError: \(underlying.localizedDescription)",
            ]
        case .wifiStatusUnknownError, .unknownError, .sessionError:
            return [
                "status": "unknown",
                "rawCode": error.code,
                "rawMessage": error.description,
            ]
        case .threadStatusError, .threadStatusDettached,
             .threadDatasetInvalid, .threadStatusNetworkNotFound,
             .threadStatusUnknownError:
            // Thread provisioning isn't part of the Dart surface yet; if a
            // Thread-only firmware response leaks through we mark it
            // unknown so callers don't misinterpret it as a Wi-Fi failure.
            return [
                "status": "unknown",
                "rawCode": error.code,
                "rawMessage": "thread error: \(error.description)",
            ]
        case .wifiStatusDisconnected:
            // Shouldn't reach here — the bridge surfaces this as an
            // exception. Map to a result anyway as a defensive fallback.
            return [
                "status": "deviceInternalError",
                "rawCode": error.code,
                "rawMessage": error.description,
            ]
        }
    }
}

File: android/src/main/kotlin/com/rainybit/esp_provisioning_flutter/ErrorMapping.kt
package com.rainybit.esp_provisioning_flutter

import com.espressif.provisioning.ESPConstants

/**
 * Tuple representing a FlutterError before it is materialised on the
 * method-channel result. Carries the same three positional fields that
 * `MethodChannel.Result.error(code, message, details)` takes, so call
 * sites can hand it straight through after marshalling to the main
 * thread.
 */
internal data class FlutterErrorTuple(
    val code: String,
    val message: String?,
    val details: Any?
)

/**
 * Translates esp-idf-provisioning-android's error surface into the typed
 * FlutterError vocabulary expected by the Dart-side `mapPlatformException`.
 *
 * Codes mirror the Swift `ErrorMapping` verbatim — the contract is the
 * sealed Dart `EspProvisioningException` hierarchy, and both native sides
 * route through the same set of strings:
 *
 *   - ble_unavailable          → BleUnavailableException
 *   - permission_denied        → PermissionDeniedException (details.permission)
 *   - device_not_found         → DeviceNotFoundException   (details.deviceId)
 *   - pop_invalid              → PopInvalidException
 *   - session_failed           → SessionFailedException
 *   - wifi_provisioning_failed → WifiProvisioningFailedException
 *   - softap_connection_failed → SoftApConnectionException (details.ssid)
 *
 * Unknown SDK exceptions fall through to `session_failed` so the Dart
 * side never receives a raw stacktrace leak.
 */
internal object ErrorMapping {

    /**
     * Maps a connect-time exception (delivered via DeviceConnectionEvent's
     * payload, or from a synchronous SDK throw) into a FlutterErrorTuple.
     *
     * Behaviour is security-aware: a generic session failure under
     * SECURITY_1 / SECURITY_2 is almost always a PoP mismatch detected
     * during the SRP6a / X25519 handshake, so we map it to `pop_invalid`
     * to drive the "re-prompt for PoP" UX. Under SECURITY_0 there is no
     * handshake so the same exception means a different kind of breakage
     * and falls through to `session_failed`.
     */
    fun connectError(
        cause: Throwable?,
        security: ESPConstants.SecurityType,
        @Suppress("UNUSED_PARAMETER") deviceId: String
    ): FlutterErrorTuple {
        val message = cause?.message
            ?: cause?.javaClass?.simpleName
            ?: "Device connection failed"
        return when (security) {
            ESPConstants.SecurityType.SECURITY_1,
            ESPConstants.SecurityType.SECURITY_2 -> FlutterErrorTuple(
                "pop_invalid",
                "Secure session handshake failed — typically a wrong PoP. ($message)",
                null
            )
            else -> FlutterErrorTuple("session_failed", message, null)
        }
    }

    /** Connect failed before the secure session was established but with
     *  no exception payload from the SDK — surfaces as session_failed
     *  irrespective of security level. */
    fun connectFailedWithoutCause(security: ESPConstants.SecurityType): FlutterErrorTuple {
        return connectError(null, security, "")
    }

    /** Unsolicited transport teardown after a session was already
     *  established. No `pop_invalid` ambiguity — the PoP was accepted
     *  earlier, so the failure is by definition a session-layer
     *  breakdown. */
    fun unsolicitedDisconnect(message: String?): FlutterErrorTuple {
        return FlutterErrorTuple(
            "session_failed",
            message ?: "Device disconnected unexpectedly",
            null
        )
    }

    /** Wi-Fi scan failures are non-security-sensitive; map all to
     *  `session_failed` so the caller can retry without re-prompting
     *  credentials. */
    fun wifiScanError(cause: Throwable?): FlutterErrorTuple {
        return FlutterErrorTuple(
            "session_failed",
            cause?.message ?: "Wi-Fi scan failed",
            null
        )
    }

    /** Maps a [ESPConstants.ProvisionFailureReason] to a `ProvisioningResult`
     *  map delivered in-band to the Dart side. AUTH_FAILED /
     *  NETWORK_NOT_FOUND are NOT exceptions — the user remedies them by
     *  re-entering input. */
    fun provisioningResultMap(reason: ESPConstants.ProvisionFailureReason): Map<String, Any?> {
        return when (reason) {
            ESPConstants.ProvisionFailureReason.AUTH_FAILED -> mapOf(
                "status" to "authFailed",
                "rawCode" to 1,
                "rawMessage" to "Device rejected the supplied passphrase"
            )
            ESPConstants.ProvisionFailureReason.NETWORK_NOT_FOUND -> mapOf(
                "status" to "networkNotFound",
                "rawCode" to 2,
                "rawMessage" to "Device could not see the supplied SSID"
            )
            ESPConstants.ProvisionFailureReason.DEVICE_DISCONNECTED -> mapOf(
                "status" to "deviceInternalError",
                "rawCode" to 3,
                "rawMessage" to "Device disconnected during credential apply"
            )
            ESPConstants.ProvisionFailureReason.UNKNOWN -> mapOf(
                "status" to "unknown",
                "rawCode" to 99,
                "rawMessage" to "Device reported an unknown provisioning failure"
            )
        }
    }

    /** Maps an arbitrary `provision()` exception (network drop, encryption
     *  failure, etc.) to a transport-level FlutterErrorTuple. Unlike
     *  [provisioningResultMap] this returns an exception envelope — the
     *  channel between phone and device collapsed before the device
     *  could report a verdict. */
    fun provisioningChannelError(cause: Throwable?): FlutterErrorTuple {
        return FlutterErrorTuple(
            "wifi_provisioning_failed",
            cause?.message ?: "Provisioning channel collapsed before device reported verdict",
            null
        )
    }

    /** Maps a sendDataToCustomEndPoint failure into a `session_failed`
     *  FlutterErrorTuple — if the custom endpoint fails the underlying
     *  secure channel has problems. */
    fun customDataError(cause: Throwable?): FlutterErrorTuple {
        return FlutterErrorTuple(
            "session_failed",
            cause?.message ?: "sendDataToCustomEndPoint failed",
            null
        )
    }
}
